### ISSUE TITLE
Support for treating backend_address_pool, http_listener, probe, backend_http_settings and request_routing_rule properties of azurerm_application_gateway as differente resources

### ISSUE THREAD

**Issue by gabrieligbastos** [hashicorp/terraform-provider-azurerm#19149]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I'm building our infra based on azure's terraform's provider and doing very well.
Got stuck on an unnecessary workload problem that could be simplified for better understandability and maintainability of the code.

The resource named:
azurerm_application_gateway

has as properties:  backend_address_pool, http_listener, probe, backend_http_settings and request_routing_rule

But when creating a new service on our side, we need to copy and past each of this property inside the application_gateway resource, just changing the variables: i.e.

 ```
resource "azurerm_application_gateway" "main" {

/* other properties and setup */
/* than for each service i need to create those 5 properties */
/* actually ten because has prod and staging slot */

backend_address_pool {
      name                              = var.appServiceHelloWorldDotnet
      fqdns                             = ["${azurerm_app_service.hello_world_sample.name}.azurewebsites.net"]
  }
 
  http_listener {
    name                                = var.appServiceHelloWorldDotnet
    frontend_ip_configuration_name      = var.appGatewayFrontendConfigurationName
    frontend_port_name                  = var.appGatewayFrontendPortNameHttps
    protocol                            = var.appGatewayHttpsListenerProtocol
    host_name                           = var.appServiceHelloWorldDotnetHostname #change
    ssl_certificate_name                = var.appGatewayCertificateName
  }
 
  probe {
    name                                = var.appServiceHelloWorldDotnet #change
    protocol                            = var.appGatewayProbeProtocol
    path                                = var.appServiceHelloWorldDotnetHealthPath #change
    host                                = "${azurerm_app_service.hello_world_sample.name}.azurewebsites.net"
    interval                            = var.appGatewayProbeInterval
    timeout                             = var.appGatewayProbeTimeout
    unhealthy_threshold                 = var.appGatewayProbeUnhealthyThreshold
  }
 
  backend_http_settings {
    name                                = var.appServiceHelloWorldDotnet #change 
    cookie_based_affinity               = var.appGatewayHttpSettingsCookieBasedAffinity
    port                                = var.appGatewayHttpSettingsPort #change
    protocol                            = var.appGatewayHttpSettingsProtocol #change
    request_timeout                     = var.appGatewayHttpSettingsRequestTimeout #change
    probe_name                          = var.appServiceHelloWorldDotnet #change
    pick_host_name_from_backend_address = var.appGatewayHttpSettingsPickHostFromBackendAddress #change
  }
 
  request_routing_rule {
    name                                = var.appServiceHelloWorldDotnet #change
    rule_type                           = var.appGatewayRequestingRouteRuleType #change
    http_listener_name                  = var.appServiceHelloWorldDotnet #change
    backend_address_pool_name           = var.appServiceHelloWorldDotnet #change
    backend_http_settings_name          = var.appServiceHelloWorldDotnet #change
  }

```

So , if those services were splited, if we had:

 ```
azurerm_application_gateway_backend_address_pool {
  application_gateway_id = {ref_to_app_gtw_id}
  /* other properties specific */
}

azurerm_application_gateway_http_listener {
  application_gateway_id = {ref_to_app_gtw_id}
  /* other properties specific */
}

azurerm_application_gateway_probe {
  application_gateway_id = {ref_to_app_gtw_id}
  /* other properties specific */
}

azurerm_application_gateway_backend_http_settings {
  application_gateway_id = {ref_to_app_gtw_id}
  /* other properties specific */
}

azurerm_application_gateway_request_routing_rule {
  application_gateway_id = {ref_to_app_gtw_id}
  /* other properties specific */
}
 ```

I would be able to add those splitted resources inside other file, and better, inside my module, so it would be really easy to replicate it to each service.

### New or Affected Resource(s)/Data Source(s)

application_gateway

### Potential Terraform Configuration

```hcl
resource "azurerm_application_gateway" "main" {
/* properties setup
}

azurerm_application_gateway_backend_address_pool {
  application_gateway_id = azurerm_application_gateway.main.id
  /* other properties specific */
}

azurerm_application_gateway_http_listener {
  application_gateway_id = azurerm_application_gateway.main.id
  /* other properties specific */
}

azurerm_application_gateway_probe {
  application_gateway_id = azurerm_application_gateway.main.id
  /* other properties specific */
}

azurerm_application_gateway_backend_http_settings {
  application_gateway_id = azurerm_application_gateway.main.id
  /* other properties specific */
}

azurerm_application_gateway_request_routing_rule {
  application_gateway_id = azurerm_application_gateway.main.id
  /* other properties specific */
}
 ```
```


### References

Imagine this: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/application_gateway

But would split those properties as mentioned above

**Comment by tombuildsstuff** [comment]
hi @gabrieligbastos 


Thanks for opening this issue.

When provisioning an Application Gateway unfortunately it's required to create at least one of most of the sub-components, as such it's not possible to add support for these as separate resources without introducing a perectual diff on the main `azurerm_application_gateway` resource. This has come up a number of times over the years and is unfortunately blocked on the Azure Service Team who need to add support for this - which is currently being tracked [in this upstream issue](https://github.com/Azure/azure-rest-api-specs/issues/2313).

Unfortunately the Azure Networking team have since tried to move this Feature Request to the Azure Feedback site, which has since been shut down - as such [we've previously asked for an update](https://github.com/Azure/azure-rest-api-specs/issues/2313#issuecomment-997844834) to get this re-opened - however it may also be worth opening a new issue on that repository too.

Since the API doesn't allow provisioning an Application Gateway without any subcomponents at the moment unfortunately and so is a feature request for the Azure Service Team - I'm going to close this issue for the moment, but I'd recommend either commenting on the existing issue above, or making a new one so that the Service Team can take a look - but unfortunately there's not much we can do from Terraform's side at the moment.

Thanks!
